File: go/internal/e2e/multi_client_test.go
package e2e

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestMultiClient(t *testing.T) {
	// Use TCP mode so a second client can connect to the same CLI process
	ctx := testharness.NewTestContext(t)
	client1 := copilot.NewClient(&copilot.ClientOptions{
		CLIPath:  ctx.CLIPath,
		Cwd:      ctx.WorkDir,
		Env:      ctx.Env(),
		UseStdio: copilot.Bool(false),
	})
	t.Cleanup(func() { client1.ForceStop() })

	// Trigger connection so we can read the port
	initSession, err := client1.CreateSession(t.Context(), &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
	})
	if err != nil {
		t.Fatalf("Failed to create init session: %v", err)
	}
	initSession.Disconnect()

	actualPort := client1.ActualPort()
	if actualPort == 0 {
		t.Fatalf("Expected non-zero port from TCP mode client")
	}

	client2 := copilot.NewClient(&copilot.ClientOptions{
		CLIUrl: fmt.Sprintf("localhost:%d", actualPort),
	})
	t.Cleanup(func() { client2.ForceStop() })

	t.Run("both clients see tool request and completion events", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		type SeedParams struct {
			Seed string `json:"seed" jsonschema:"A seed value"`
		}

		tool := copilot.DefineTool("magic_number", "Returns a magic number",
			func(params SeedParams, inv copilot.ToolInvocation) (string, error) {
				return fmt.Sprintf("MAGIC_%s_42", params.Seed), nil
			})

		// Client 1 creates a session with a custom tool
		session1, err := client1.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools:               []copilot.Tool{tool},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Client 2 resumes with NO tools — should not overwrite client 1's tools
		session2, err := client2.ResumeSession(t.Context(), session1.SessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		// Set up event waiters BEFORE sending the prompt to avoid race conditions
		client1Requested := make(chan struct{}, 1)
		client2Requested := make(chan struct{}, 1)
		client1Completed := make(chan struct{}, 1)
		client2Completed := make(chan struct{}, 1)

		session1.On(func(event copilot.SessionEvent) {
			if event.Type == copilot.SessionEventTypeExternalToolRequested {
				select {
				case client1Requested <- struct{}{}:
				default:
				}
			}
			if event.Type == copilot.SessionEventTypeExternalToolCompleted {
				select {
				case client1Completed <- struct{}{}:
				default:
				}
			}
		})
		session2.On(func(event copilot.SessionEvent) {
			if event.Type == copilot.SessionEventTypeExternalToolRequested {
				select {
				case client2Requested <- struct{}{}:
				default:
				}
			}
			if event.Type == copilot.SessionEventTypeExternalToolCompleted {
				select {
				case client2Completed <- struct{}{}:
				default:
				}
			}
		})

		// Send a prompt that triggers the custom tool
		response, err := session1.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Use the magic_number tool with seed 'hello' and tell me the result",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		if response == nil || response.Data.Content == nil || !strings.Contains(*response.Data.Content, "MAGIC_hello_42") {
			t.Errorf("Expected response to contain 'MAGIC_hello_42', got %v", response)
		}

		// Wait for all broadcast events to arrive on both clients
		timeout := time.After(30 * time.Second)
		for _, ch := range []chan struct{}{client1Requested, client2Requested, client1Completed, client2Completed} {
			select {
			case <-ch:
			case <-timeout:
				t.Fatal("Timed out waiting for broadcast events on both clients")
			}
		}

		session2.Disconnect()
	})

	t.Run("one client approves permission and both see the result", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		var client1PermissionRequests []copilot.PermissionRequest
		var mu sync.Mutex

		// Client 1 creates a session and manually approves permission requests
		session1, err := client1.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: func(request copilot.PermissionRequest, invocation copilot.PermissionInvocation) (copilot.PermissionRequestResult, error) {
				mu.Lock()
				client1PermissionRequests = append(client1PermissionRequests, request)
				mu.Unlock()
				return copilot.PermissionRequestResult{Kind: copilot.PermissionRequestResultKindApproved}, nil
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Client 2 resumes — its handler never resolves, so only client 1's approval takes effect
		session2, err := client2.ResumeSession(t.Context(), session1.SessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: func(request copilot.PermissionRequest, invocation copilot.PermissionInvocation) (copilot.PermissionRequestResult, error) {
				// Block forever so only client 1's handler responds
				select {}
			},
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		// Track events
		var client1Events, client2Events []copilot.SessionEvent
		var mu1, mu2 sync.Mutex
		session1.On(func(event copilot.SessionEvent) {
			mu1.Lock()
			client1Events = append(client1Events, event)
			mu1.Unlock()
		})
		session2.On(func(event copilot.SessionEvent) {
			mu2.Lock()
			client2Events = append(client2Events, event)
			mu2.Unlock()
		})

		// Send a prompt that triggers a write operation (requires permission)
		response, err := session1.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Create a file called hello.txt containing the text 'hello world'",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if response == nil || response.Data.Content == nil || *response.Data.Content == "" {
			t.Errorf("Expected non-empty response")
		}

		// Client 1 should have handled the permission request
		mu.Lock()
		permCount := len(client1PermissionRequests)
		mu.Unlock()
		if permCount == 0 {
			t.Errorf("Expected client 1 to handle at least one permission request")
		}

		// Both clients should have seen permission.requested events
		mu1.Lock()
		c1PermRequested := filterEventsByType(client1Events, copilot.SessionEventTypePermissionRequested)
		mu1.Unlock()
		mu2.Lock()
		c2PermRequested := filterEventsByType(client2Events, copilot.SessionEventTypePermissionRequested)
		mu2.Unlock()

		if len(c1PermRequested) == 0 {
			t.Errorf("Expected client 1 to see permission.requested events")
		}
		if len(c2PermRequested) == 0 {
			t.Errorf("Expected client 2 to see permission.requested events")
		}

		// Both clients should have seen permission.completed events with approved result
		mu1.Lock()
		c1PermCompleted := filterEventsByType(client1Events, copilot.SessionEventTypePermissionCompleted)
		mu1.Unlock()
		mu2.Lock()
		c2PermCompleted := filterEventsByType(client2Events, copilot.SessionEventTypePermissionCompleted)
		mu2.Unlock()

		if len(c1PermCompleted) == 0 {
			t.Errorf("Expected client 1 to see permission.completed events")
		}
		if len(c2PermCompleted) == 0 {
			t.Errorf("Expected client 2 to see permission.completed events")
		}
		for _, event := range append(c1PermCompleted, c2PermCompleted...) {
			if event.Data.Result == nil || event.Data.Result.Kind == nil || *event.Data.Result.Kind != "approved" {
				t.Errorf("Expected permission.completed result kind 'approved', got %v", event.Data.Result)
			}
		}

		session2.Disconnect()
	})

	t.Run("one client rejects permission and both see the result", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		// Client 1 creates a session and denies all permission requests
		session1, err := client1.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: func(request copilot.PermissionRequest, invocation copilot.PermissionInvocation) (copilot.PermissionRequestResult, error) {
				return copilot.PermissionRequestResult{Kind: copilot.PermissionRequestResultKindDeniedInteractivelyByUser}, nil
			},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Client 2 resumes — its handler never resolves so only client 1's denial takes effect
		session2, err := client2.ResumeSession(t.Context(), session1.SessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: func(request copilot.PermissionRequest, invocation copilot.PermissionInvocation) (copilot.PermissionRequestResult, error) {
				select {}
			},
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		var client1Events, client2Events []copilot.SessionEvent
		var mu1, mu2 sync.Mutex
		session1.On(func(event copilot.SessionEvent) {
			mu1.Lock()
			client1Events = append(client1Events, event)
			mu1.Unlock()
		})
		session2.On(func(event copilot.SessionEvent) {
			mu2.Lock()
			client2Events = append(client2Events, event)
			mu2.Unlock()
		})

		// Write a test file and ask the agent to edit it
		testFile := filepath.Join(ctx.WorkDir, "protected.txt")
		if err := os.WriteFile(testFile, []byte("protected content"), 0644); err != nil {
			t.Fatalf("Failed to write test file: %v", err)
		}

		_, err = session1.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Edit protected.txt and replace 'protected' with 'hacked'.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}

		// Verify the file was NOT modified (permission was denied)
		content, err := os.ReadFile(testFile)
		if err != nil {
			t.Fatalf("Failed to read test file: %v", err)
		}
		if string(content) != "protected content" {
			t.Errorf("Expected file content 'protected content', got '%s'", string(content))
		}

		// Both clients should have seen permission.requested events
		mu1.Lock()
		c1PermRequested := filterEventsByType(client1Events, copilot.SessionEventTypePermissionRequested)
		mu1.Unlock()
		mu2.Lock()
		c2PermRequested := filterEventsByType(client2Events, copilot.SessionEventTypePermissionRequested)
		mu2.Unlock()

		if len(c1PermRequested) == 0 {
			t.Errorf("Expected client 1 to see permission.requested events")
		}
		if len(c2PermRequested) == 0 {
			t.Errorf("Expected client 2 to see permission.requested events")
		}

		// Both clients should see the denial in the completed event
		mu1.Lock()
		c1PermCompleted := filterEventsByType(client1Events, copilot.SessionEventTypePermissionCompleted)
		mu1.Unlock()
		mu2.Lock()
		c2PermCompleted := filterEventsByType(client2Events, copilot.SessionEventTypePermissionCompleted)
		mu2.Unlock()

		if len(c1PermCompleted) == 0 {
			t.Errorf("Expected client 1 to see permission.completed events")
		}
		if len(c2PermCompleted) == 0 {
			t.Errorf("Expected client 2 to see permission.completed events")
		}
		for _, event := range append(c1PermCompleted, c2PermCompleted...) {
			if event.Data.Result == nil || event.Data.Result.Kind == nil || *event.Data.Result.Kind != "denied-interactively-by-user" {
				t.Errorf("Expected permission.completed result kind 'denied-interactively-by-user', got %v", event.Data.Result)
			}
		}

		session2.Disconnect()
	})

	t.Run("two clients register different tools and agent uses both", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		type CountryCodeParams struct {
			CountryCode string `json:"countryCode" jsonschema:"A two-letter country code"`
		}

		toolA := copilot.DefineTool("city_lookup", "Returns a city name for a given country code",
			func(params CountryCodeParams, inv copilot.ToolInvocation) (string, error) {
				return fmt.Sprintf("CITY_FOR_%s", params.CountryCode), nil
			})

		toolB := copilot.DefineTool("currency_lookup", "Returns a currency for a given country code",
			func(params CountryCodeParams, inv copilot.ToolInvocation) (string, error) {
				return fmt.Sprintf("CURRENCY_FOR_%s", params.CountryCode), nil
			})

		// Client 1 creates a session with tool A
		session1, err := client1.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools:               []copilot.Tool{toolA},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Client 2 resumes with tool B (different tool, union should have both)
		session2, err := client2.ResumeSession(t.Context(), session1.SessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools:               []copilot.Tool{toolB},
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		// Send prompts sequentially to avoid nondeterministic tool_call ordering
		response1, err := session1.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Use the city_lookup tool with countryCode 'US' and tell me the result.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if response1 == nil || response1.Data.Content == nil {
			t.Fatalf("Expected response with content")
		}
		if !strings.Contains(*response1.Data.Content, "CITY_FOR_US") {
			t.Errorf("Expected response to contain 'CITY_FOR_US', got '%s'", *response1.Data.Content)
		}

		response2, err := session1.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Now use the currency_lookup tool with countryCode 'US' and tell me the result.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if response2 == nil || response2.Data.Content == nil {
			t.Fatalf("Expected response with content")
		}
		if !strings.Contains(*response2.Data.Content, "CURRENCY_FOR_US") {
			t.Errorf("Expected response to contain 'CURRENCY_FOR_US', got '%s'", *response2.Data.Content)
		}

		session2.Disconnect()
	})

	t.Run("disconnecting client removes its tools", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		type InputParams struct {
			Input string `json:"input" jsonschema:"Input string"`
		}

		toolA := copilot.DefineTool("stable_tool", "A tool that persists across disconnects",
			func(params InputParams, inv copilot.ToolInvocation) (string, error) {
				return fmt.Sprintf("STABLE_%s", params.Input), nil
			})

		toolB := copilot.DefineTool("ephemeral_tool", "A tool that will disappear when its client disconnects",
			func(params InputParams, inv copilot.ToolInvocation) (string, error) {
				return fmt.Sprintf("EPHEMERAL_%s", params.Input), nil
			})

		// Client 1 creates a session with stable_tool
		session1, err := client1.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools:               []copilot.Tool{toolA},
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}

		// Client 2 resumes with ephemeral_tool
		_, err = client2.ResumeSession(t.Context(), session1.SessionID, &copilot.ResumeSessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
			Tools:               []copilot.Tool{toolB},
		})
		if err != nil {
			t.Fatalf("Failed to resume session: %v", err)
		}

		// Verify both tools work before disconnect (sequential to avoid nondeterministic tool_call ordering)
		stableResponse, err := session1.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Use the stable_tool with input 'test1' and tell me the result.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if stableResponse == nil || stableResponse.Data.Content == nil {
			t.Fatalf("Expected response with content")
		}
		if !strings.Contains(*stableResponse.Data.Content, "STABLE_test1") {
			t.Errorf("Expected response to contain 'STABLE_test1', got '%s'", *stableResponse.Data.Content)
		}

		ephemeralResponse, err := session1.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Use the ephemeral_tool with input 'test2' and tell me the result.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if ephemeralResponse == nil || ephemeralResponse.Data.Content == nil {
			t.Fatalf("Expected response with content")
		}
		if !strings.Contains(*ephemeralResponse.Data.Content, "EPHEMERAL_test2") {
			t.Errorf("Expected response to contain 'EPHEMERAL_test2', got '%s'", *ephemeralResponse.Data.Content)
		}

		// Disconnect client 2 without destroying the shared session
		client2.ForceStop()

		// Give the server time to process the connection close and remove tools
		time.Sleep(500 * time.Millisecond)

		// Recreate client2 for cleanup (but don't rejoin the session)
		client2 = copilot.NewClient(&copilot.ClientOptions{
			CLIUrl: fmt.Sprintf("localhost:%d", actualPort),
		})

		// Now only stable_tool should be available
		afterResponse, err := session1.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Use the stable_tool with input 'still_here'. Also try using ephemeral_tool if it is available.",
		})
		if err != nil {
			t.Fatalf("Failed to send message: %v", err)
		}
		if afterResponse == nil || afterResponse.Data.Content == nil {
			t.Fatalf("Expected response with content")
		}
		if !strings.Contains(*afterResponse.Data.Content, "STABLE_still_here") {
			t.Errorf("Expected response to contain 'STABLE_still_here', got '%s'", *afterResponse.Data.Content)
		}
		// ephemeral_tool should NOT have produced a result
		if strings.Contains(*afterResponse.Data.Content, "EPHEMERAL_") {
			t.Errorf("Expected response NOT to contain 'EPHEMERAL_', got '%s'", *afterResponse.Data.Content)
		}
	})
}

func filterEventsByType(events []copilot.SessionEvent, eventType copilot.SessionEventType) []copilot.SessionEvent {
	var filtered []copilot.SessionEvent
	for _, e := range events {
		if e.Type == eventType {
			filtered = append(filtered, e)
		}
	}
	return filtered
}

File: nodejs/test/e2e/multi-client.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { describe, expect, it, afterAll } from "vitest";
import { z } from "zod";
import { CopilotClient, defineTool, approveAll } from "../../src/index.js";
import type { SessionEvent } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext";

describe("Multi-client broadcast", async () => {
    // Use TCP mode so a second client can connect to the same CLI process
    const ctx = await createSdkTestContext({ useStdio: false });
    const client1 = ctx.copilotClient;

    // Trigger connection so we can read the port
    const initSession = await client1.createSession({ onPermissionRequest: approveAll });
    await initSession.disconnect();

    const actualPort = (client1 as unknown as { actualPort: number }).actualPort;
    let client2 = new CopilotClient({ cliUrl: `localhost:${actualPort}` });

    afterAll(async () => {
        await client2.stop();
    });

    it("both clients see tool request and completion events", async () => {
        const tool = defineTool("magic_number", {
            description: "Returns a magic number",
            parameters: z.object({
                seed: z.string().describe("A seed value"),
            }),
            handler: ({ seed }) => `MAGIC_${seed}_42`,
        });

        // Client 1 creates a session with a custom tool
        const session1 = await client1.createSession({
            onPermissionRequest: approveAll,
            tools: [tool],
        });

        // Client 2 resumes with NO tools — should not overwrite client 1's tools
        const session2 = await client2.resumeSession(session1.sessionId, {
            onPermissionRequest: approveAll,
        });

        // Set up event waiters BEFORE sending the prompt to avoid race conditions
        const waitForEvent = (session: typeof session1, type: string) =>
            new Promise<SessionEvent>((resolve) => {
                const unsub = session.on((event) => {
                    if (event.type === type) {
                        unsub();
                        resolve(event);
                    }
                });
            });

        const client1RequestedP = waitForEvent(session1, "external_tool.requested");
        const client2RequestedP = waitForEvent(session2, "external_tool.requested");
        const client1CompletedP = waitForEvent(session1, "external_tool.completed");
        const client2CompletedP = waitForEvent(session2, "external_tool.completed");

        // Send a prompt that triggers the custom tool
        const response = await session1.sendAndWait({
            prompt: "Use the magic_number tool with seed 'hello' and tell me the result",
        });

        // The response should contain the tool's output
        expect(response?.data.content).toContain("MAGIC_hello_42");

        // Wait for all broadcast events to arrive on both clients
        await expect(
            Promise.all([
                client1RequestedP,
                client2RequestedP,
                client1CompletedP,
                client2CompletedP,
            ])
        ).resolves.toBeDefined();

        await session2.disconnect();
    });

    it("one client approves permission and both see the result", async () => {
        const client1PermissionRequests: unknown[] = [];

        // Client 1 creates a session and manually approves permission requests
        const session1 = await client1.createSession({
            onPermissionRequest: (request) => {
                client1PermissionRequests.push(request);
                return { kind: "approve-once" as const };
            },
        });

        // Client 2 resumes the same session — its handler never resolves,
        // so only client 1's approval takes effect (no race)
        const session2 = await client2.resumeSession(session1.sessionId, {
            onPermissionRequest: () => new Promise(() => {}),
        });

        // Track events seen by each client
        const client1Events: SessionEvent[] = [];
        const client2Events: SessionEvent[] = [];

        session1.on((event) => client1Events.push(event));
        session2.on((event) => client2Events.push(event));

        // Send a prompt that triggers a write operation (requires permission)
        const response = await session1.sendAndWait({
            prompt: "Create a file called hello.txt containing the text 'hello world'",
        });

        expect(response?.data.content).toBeTruthy();

        // Client 1 should have handled the permission request
        expect(client1PermissionRequests.length).toBeGreaterThan(0);

        // Both clients should have seen permission.requested events
        const client1PermRequested = client1Events.filter((e) => e.type === "permission.requested");
        const client2PermRequested = client2Events.filter((e) => e.type === "permission.requested");
        expect(client1PermRequested.length).toBeGreaterThan(0);
        expect(client2PermRequested.length).toBeGreaterThan(0);

        // Both clients should have seen permission.completed events with approved result
        const client1PermCompleted = client1Events.filter(
            (e): e is SessionEvent & { type: "permission.completed" } =>
                e.type === "permission.completed"
        );
        const client2PermCompleted = client2Events.filter(
            (e): e is SessionEvent & { type: "permission.completed" } =>
                e.type === "permission.completed"
        );
        expect(client1PermCompleted.length).toBeGreaterThan(0);
        expect(client2PermCompleted.length).toBeGreaterThan(0);
        for (const event of [...client1PermCompleted, ...client2PermCompleted]) {
            expect(event.data.result.kind).toBe("approved");
        }

        await session2.disconnect();
    });

    it("one client rejects permission and both see the result", async () => {
        // Client 1 creates a session and denies all permission requests
        const session1 = await client1.createSession({
            onPermissionRequest: () => ({ kind: "reject" as const }),
        });

        // Client 2 resumes — its handler never resolves so only client 1's denial takes effect
        const session2 = await client2.resumeSession(session1.sessionId, {
            onPermissionRequest: () => new Promise(() => {}),
        });

        const client1Events: SessionEvent[] = [];
        const client2Events: SessionEvent[] = [];

        session1.on((event) => client1Events.push(event));
        session2.on((event) => client2Events.push(event));

        // Ask the agent to write a file (requires permission)
        const { writeFile } = await import("fs/promises");
        const { join } = await import("path");
        const testFile = join(ctx.workDir, "protected.txt");
        await writeFile(testFile, "protected content");

        await session1.sendAndWait({
            prompt: "Edit protected.txt and replace 'protected' with 'hacked'.",
        });

        // Verify the file was NOT modified (permission was denied)
        const { readFile } = await import("fs/promises");
        const content = await readFile(testFile, "utf-8");
        expect(content).toBe("protected content");

        // Both clients should have seen permission.requested and permission.completed
        expect(
            client1Events.filter((e) => e.type === "permission.requested").length
        ).toBeGreaterThan(0);
        expect(
            client2Events.filter((e) => e.type === "permission.requested").length
        ).toBeGreaterThan(0);

        // Both clients should see the denial in the completed event
        const client1PermCompleted = client1Events.filter(
            (e): e is SessionEvent & { type: "permission.completed" } =>
                e.type === "permission.completed"
        );
        const client2PermCompleted = client2Events.filter(
            (e): e is SessionEvent & { type: "permission.completed" } =>
                e.type === "permission.completed"
        );
        expect(client1PermCompleted.length).toBeGreaterThan(0);
        expect(client2PermCompleted.length).toBeGreaterThan(0);
        for (const event of [...client1PermCompleted, ...client2PermCompleted]) {
            expect(event.data.result.kind).toBe("denied-interactively-by-user");
        }

        await session2.disconnect();
    });

    it(
        "two clients register different tools and agent uses both",
        { timeout: 90_000 },
        async () => {
            const toolA = defineTool("city_lookup", {
                description: "Returns a city name for a given country code",
                parameters: z.object({
                    countryCode: z.string().describe("A two-letter country code"),
                }),
                handler: ({ countryCode }) => `CITY_FOR_${countryCode}`,
            });

            const toolB = defineTool("currency_lookup", {
                description: "Returns a currency for a given country code",
                parameters: z.object({
                    countryCode: z.string().describe("A two-letter country code"),
                }),
                handler: ({ countryCode }) => `CURRENCY_FOR_${countryCode}`,
            });

            // Client 1 creates a session with tool A
            const session1 = await client1.createSession({
                onPermissionRequest: approveAll,
                tools: [toolA],
            });

            // Client 2 resumes with tool B (different tool, union should have both)
            const session2 = await client2.resumeSession(session1.sessionId, {
                onPermissionRequest: approveAll,
                tools: [toolB],
            });

            // Send prompts sequentially to avoid nondeterministic tool_call ordering
            const response1 = await session1.sendAndWait({
                prompt: "Use the city_lookup tool with countryCode 'US' and tell me the result.",
            });
            expect(response1?.data.content).toContain("CITY_FOR_US");

            const response2 = await session1.sendAndWait({
                prompt: "Now use the currency_lookup tool with countryCode 'US' and tell me the result.",
            });
            expect(response2?.data.content).toContain("CURRENCY_FOR_US");

            await session2.disconnect();
        }
    );

    it("disconnecting client removes its tools", { timeout: 90_000 }, async () => {
        const toolA = defineTool("stable_tool", {
            description: "A tool that persists across disconnects",
            parameters: z.object({ input: z.string() }),
            handler: ({ input }) => `STABLE_${input}`,
        });

        const toolB = defineTool("ephemeral_tool", {
            description: "A tool that will disappear when its client disconnects",
            parameters: z.object({ input: z.string() }),
            handler: ({ input }) => `EPHEMERAL_${input}`,
        });

        // Client 1 creates a session with stable_tool
        const session1 = await client1.createSession({
            onPermissionRequest: approveAll,
            tools: [toolA],
        });

        // Client 2 resumes with ephemeral_tool
        await client2.resumeSession(session1.sessionId, {
            onPermissionRequest: approveAll,
            tools: [toolB],
        });

        // Verify both tools work before disconnect (sequential to avoid nondeterministic tool_call ordering)
        const stableResponse = await session1.sendAndWait({
            prompt: "Use the stable_tool with input 'test1' and tell me the result.",
        });
        expect(stableResponse?.data.content).toContain("STABLE_test1");

        const ephemeralResponse = await session1.sendAndWait({
            prompt: "Use the ephemeral_tool with input 'test2' and tell me the result.",
        });
        expect(ephemeralResponse?.data.content).toContain("EPHEMERAL_test2");

        // Disconnect client 2 without destroying the shared session.
        // Suppress "Connection is disposed" rejections that occur when the server
        // broadcasts events (e.g. tool_changed_notice) to the now-dead connection.
        const suppressDisposed = (reason: unknown) => {
            if (reason instanceof Error && reason.message.includes("Connection is disposed")) {
                return;
            }
            throw reason;
        };
        process.on("unhandledRejection", suppressDisposed);
        await client2.forceStop();

        // Give the server time to process the connection close and remove tools
        await new Promise((resolve) => setTimeout(resolve, 500));
        process.removeListener("unhandledRejection", suppressDisposed);

        // Recreate client2 for cleanup in afterAll (but don't rejoin the session)
        client2 = new CopilotClient({ cliUrl: `localhost:${actualPort}` });

        // Now only stable_tool should be available
        const afterResponse = await session1.sendAndWait({
            prompt: "Use the stable_tool with input 'still_here'. Also try using ephemeral_tool if it is available.",
        });
        expect(afterResponse?.data.content).toContain("STABLE_still_here");
        // ephemeral_tool should NOT have produced a result
        expect(afterResponse?.data.content).not.toContain("EPHEMERAL_");
    });
});
